File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/bridge/text/TextBridge.kt
package dev.destack.runtime.android.bridge.text

import dev.destack.runtime.android.core.HostSessionHandle
import dev.destack.runtime.android.core.RuntimeHost
import dev.destack.runtime.android.core.hostStatusOk
import dev.destack.runtime.android.input.text.RuntimeHostTextInputCloseRequest
import dev.destack.runtime.android.input.text.RuntimeHostTextInputConfiguration
import dev.destack.runtime.android.input.text.RuntimeHostTextInputEvent
import dev.destack.runtime.android.input.text.RuntimeHostTextInputGeometry
import dev.destack.runtime.android.input.text.RuntimeHostTextInputGeometryRequest
import dev.destack.runtime.android.input.text.RuntimeHostTextInputOpenRequest
import dev.destack.runtime.android.input.text.RuntimeHostTextInputRectangle
import dev.destack.runtime.android.input.text.RuntimeHostTextInputRange
import dev.destack.runtime.android.input.text.RuntimeHostTextInputState
import dev.destack.runtime.android.input.text.RuntimeHostTextInputStateRequest
import dev.destack.runtime.android.input.text.RuntimeHostTextInputTransform2D
import dev.destack.runtime.android.input.text.RuntimeHostTextInputType
import dev.destack.runtime.android.input.text.TextInputEvents

/**
 * One text-input bridge lane for one attached Android runtime host.
 */
internal class TextBridge(
    private val sessionHandle: HostSessionHandle,
    private val bindings: TextAbi,
) : TextInputEvents {
    /**
     * Send one host text-input event into the runtime ingress path.
     */
    override fun sendTextInputEvent(
        event: RuntimeHostTextInputEvent,
    ) {
        val status = bindings.notifyTextInputState(
            sessionHandle = sessionHandle,
            sessionId = event.sessionId,
            state = event.state,
        )

        require(status.code == hostStatusOk) {
            "runtime bridge could not deliver text-input state: code ${status.code}, error ${status.errorId}"
        }
    }

    /**
     * Open one Android text-input session through the attached host.
     */
    fun openTextInput(
        runtimeHost: RuntimeHost,
        sessionId: Long,
        inputType: Int,
        isMultiline: Boolean,
        isSecure: Boolean,
        text: String,
        selectionStart: Int,
        selectionEnd: Int,
        hasComposing: Boolean,
        composingStart: Int,
        composingEnd: Int,
    ): Int {
        return runtimeHost.textInputRequests.openTextInput(
            RuntimeHostTextInputOpenRequest(
                configuration = RuntimeHostTextInputConfiguration(
                    sessionId = sessionId,
                    inputType = decodeInputType(inputType),
                    isMultiline = isMultiline,
                    isSecure = isSecure,
                ),
                state = decodeState(
                    text = text,
                    selectionStart = selectionStart,
                    selectionEnd = selectionEnd,
                    hasComposing = hasComposing,
                    composingStart = composingStart,
                    composingEnd = composingEnd,
                ),
            ),
        )
    }

    /**
     * Close one Android text-input session through the attached host.
     */
    fun closeTextInput(
        runtimeHost: RuntimeHost,
        sessionId: Long,
    ): Int {
        return runtimeHost.textInputRequests.closeTextInput(
            RuntimeHostTextInputCloseRequest(
                sessionId = sessionId,
            ),
        )
    }

    /**
     * Update one Android text-input geometry through the attached host.
     */
    fun setTextInputGeometry(
        runtimeHost: RuntimeHost,
        sessionId: Long,
        transformXx: Double,
        transformXy: Double,
        transformYx: Double,
        transformYy: Double,
        transformTx: Double,
        transformTy: Double,
        editorX: Double,
        editorY: Double,
        editorWidth: Double,
        editorHeight: Double,
        hasCaretRectangle: Boolean,
        caretX: Double,
        caretY: Double,
        caretWidth: Double,
        caretHeight: Double,
        hasComposingRectangle: Boolean,
        composingX: Double,
        composingY: Double,
        composingWidth: Double,
        composingHeight: Double,
    ): Int {
        return runtimeHost.textInputRequests.setTextInputGeometry(
            RuntimeHostTextInputGeometryRequest(
                sessionId = sessionId,
                geometry = RuntimeHostTextInputGeometry(
                    localToTargetTransform = RuntimeHostTextInputTransform2D(
                        xx = transformXx,
                        xy = transformXy,
                        yx = transformYx,
                        yy = transformYy,
                        tx = transformTx,
                        ty = transformTy,
                    ),
                    editorRectangle = RuntimeHostTextInputRectangle(
                        x = editorX,
                        y = editorY,
                        width = editorWidth,
                        height = editorHeight,
                    ),
                    caretRectangle = if (hasCaretRectangle) {
                        RuntimeHostTextInputRectangle(
                            x = caretX,
                            y = caretY,
                            width = caretWidth,
                            height = caretHeight,
                        )
                    } else {
                        null
                    },
                    composingRectangle = if (hasComposingRectangle) {
                        RuntimeHostTextInputRectangle(
                            x = composingX,
                            y = composingY,
                            width = composingWidth,
                            height = composingHeight,
                        )
                    } else {
                        null
                    },
                ),
            ),
        )
    }

    /**
     * Update one Android text-input state through the attached host.
     */
    fun setTextInputState(
        runtimeHost: RuntimeHost,
        sessionId: Long,
        text: String,
        selectionStart: Int,
        selectionEnd: Int,
        hasComposing: Boolean,
        composingStart: Int,
        composingEnd: Int,
    ): Int {
        return runtimeHost.textInputRequests.setTextInputState(
            RuntimeHostTextInputStateRequest(
                sessionId = sessionId,
                state = decodeState(
                    text = text,
                    selectionStart = selectionStart,
                    selectionEnd = selectionEnd,
                    hasComposing = hasComposing,
                    composingStart = composingStart,
                    composingEnd = composingEnd,
                ),
            ),
        )
    }

    /**
     * Decode one bridge text-input type value.
     */
    private fun decodeInputType(
        value: Int,
    ): RuntimeHostTextInputType {
        return RuntimeHostTextInputType.entries.firstOrNull { entry ->
            entry.ordinal == value
        } ?: RuntimeHostTextInputType.TEXT
    }

    /**
     * Decode one bridge text-input state payload.
     */
    private fun decodeState(
        text: String,
        selectionStart: Int,
        selectionEnd: Int,
        hasComposing: Boolean,
        composingStart: Int,
        composingEnd: Int,
    ): RuntimeHostTextInputState {
        return RuntimeHostTextInputState(
            text = text,
            selection = RuntimeHostTextInputRange(
                startOffset = selectionStart,
                endOffset = selectionEnd,
            ),
            composing = if (hasComposing) {
                RuntimeHostTextInputRange(
                    startOffset = composingStart,
                    endOffset = composingEnd,
                )
            } else {
                null
            },
        )
    }
}

File: language/runtime/apple/ios/Sources/RuntimeHostIOS/Bridge/Text/TextBridge.swift
import Foundation
import RuntimeHostAppleBridgeC
import RuntimeHostAppleCore

let textOpenCallback: TextOpenCallback = {
  sessionHandle,
  request in
  handleTextOpen(
    sessionHandle: sessionHandle,
    request: request
  )
}

let textCloseCallback: TextCloseCallback = {
  sessionHandle,
  sessionID in
  handleTextClose(
    sessionHandle: sessionHandle,
    sessionID: sessionID
  )
}

let textSetGeometryCallback: TextGeometryCallback = {
  sessionHandle,
  request in
  handleTextSetGeometry(
    sessionHandle: sessionHandle,
    request: request
  )
}

let textSetStateCallback: TextStateCallback = {
  sessionHandle,
  request in
  handleTextSetState(
    sessionHandle: sessionHandle,
    request: request
  )
}

/// One text bridge lane for one attached iOS runtime host.
@MainActor
final class TextBridge: TextInputEvents {
  /// The runtime session routed through this text bridge lane.
  let sessionHandle: HostSessionHandle
  private let bindings: any TextAbi

  /// Create one text bridge lane.
  init(
    sessionHandle: HostSessionHandle,
    bindings: any TextAbi
  ) {
    self.sessionHandle = sessionHandle
    self.bindings = bindings
  }

  /// Send one host text-input event into the runtime ingress path.
  func sendTextInputEvent(
    _ event: RuntimeHostTextInputEvent
  ) {
    let status = bindings.notifyTextInputState(
      sessionHandle: sessionHandle,
      sessionID: event.sessionID,
      state: event.state
    )

    precondition(
      status.code == hostStatusOk,
      "runtime bridge could not deliver text-input state: code \(status.code), error \(status.errorID)"
    )
  }

  /// Open one iOS text session through the attached host.
  func openTextInput(
    runtimeHost: RuntimeHost,
    _ request: RuntimeHostTextInputOpenRequest
  ) -> UInt32 {
    runtimeHost.textInputRequests.openTextInput(request)
  }

  /// Close one iOS text session through the attached host.
  func closeTextInput(
    runtimeHost: RuntimeHost,
    _ request: RuntimeHostTextInputCloseRequest
  ) -> UInt32 {
    runtimeHost.textInputRequests.closeTextInput(request)
  }

  /// Update one iOS text geometry through the attached host.
  func setTextInputGeometry(
    runtimeHost: RuntimeHost,
    _ request: RuntimeHostTextInputGeometryRequest
  ) -> UInt32 {
    runtimeHost.textInputRequests.setTextInputGeometry(request)
  }

  /// Update one iOS text state through the attached host.
  func setTextInputState(
    runtimeHost: RuntimeHost,
    _ request: RuntimeHostTextInputStateRequest
  ) -> UInt32 {
    runtimeHost.textInputRequests.setTextInputState(request)
  }
}

/// Resolve one registered text bridge for one runtime session.
func guardTextBridge(
  sessionHandle: UInt64
) -> TextBridge? {
  RuntimeBridgeRegistry.resolve(sessionHandle: sessionHandle)?.textBridge
}

/// Handle one runtime callback asking to open one iOS text session.
private func handleTextOpen(
  sessionHandle: UInt64,
  request bridgeRequest: DestackRustTextOpenRequest
) -> UInt32 {
  guard let bridge = guardTextBridge(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }
  guard let runtimeHost = resolveRuntimeHost(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }

  let request: RuntimeHostTextInputOpenRequest
  do {
    request = try decodeTextOpenRequest(bridgeRequest)
  } catch {
    return hostStatusInvalidArgument
  }

  return runOnMainThread {
    bridge.openTextInput(
      runtimeHost: runtimeHost,
      request
    )
  }
}

/// Handle one runtime callback asking to close one iOS text session.
private func handleTextClose(
  sessionHandle: UInt64,
  sessionID: UInt64
) -> UInt32 {
  guard let bridge = guardTextBridge(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }
  guard let runtimeHost = resolveRuntimeHost(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }

  return runOnMainThread {
    bridge.closeTextInput(
      runtimeHost: runtimeHost,
      RuntimeHostTextInputCloseRequest(sessionID: sessionID)
    )
  }
}

/// Handle one runtime callback asking to update one iOS text geometry.
private func handleTextSetGeometry(
  sessionHandle: UInt64,
  request bridgeRequest: DestackRustTextGeometryRequest
) -> UInt32 {
  guard let bridge = guardTextBridge(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }
  guard let runtimeHost = resolveRuntimeHost(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }

  let request = RuntimeHostTextInputGeometryRequest(
    sessionID: bridgeRequest.session_id,
    geometry: RuntimeHostTextInputGeometry(
      localToTargetTransform: RuntimeHostTextInputTransform2D(
        xx: bridgeRequest.local_to_target_transform.xx,
        xy: bridgeRequest.local_to_target_transform.xy,
        yx: bridgeRequest.local_to_target_transform.yx,
        yy: bridgeRequest.local_to_target_transform.yy,
        tx: bridgeRequest.local_to_target_transform.tx,
        ty: bridgeRequest.local_to_target_transform.ty
      ),
      editorRectangle: RuntimeHostTextInputRectangle(
        x: bridgeRequest.editor_rectangle.x,
        y: bridgeRequest.editor_rectangle.y,
        width: bridgeRequest.editor_rectangle.width,
        height: bridgeRequest.editor_rectangle.height
      ),
      caretRectangle: bridgeRequest.has_caret_rectangle
        ? RuntimeHostTextInputRectangle(
          x: bridgeRequest.caret_rectangle.x,
          y: bridgeRequest.caret_rectangle.y,
          width: bridgeRequest.caret_rectangle.width,
          height: bridgeRequest.caret_rectangle.height
        )
        : nil,
      composingRectangle: bridgeRequest.has_composing_rectangle
        ? RuntimeHostTextInputRectangle(
          x: bridgeRequest.composing_rectangle.x,
          y: bridgeRequest.composing_rectangle.y,
          width: bridgeRequest.composing_rectangle.width,
          height: bridgeRequest.composing_rectangle.height
        )
        : nil
    )
  )

  return runOnMainThread {
    bridge.setTextInputGeometry(
      runtimeHost: runtimeHost,
      request
    )
  }
}

/// Handle one runtime callback asking to update one iOS text state.
private func handleTextSetState(
  sessionHandle: UInt64,
  request bridgeRequest: DestackRustTextStateRequest
) -> UInt32 {
  guard let bridge = guardTextBridge(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }
  guard let runtimeHost = resolveRuntimeHost(sessionHandle: sessionHandle) else {
    return hostStatusNotFound
  }

  let state: RuntimeHostTextInputState
  do {
    state = try decodeTextState(bridgeRequest.state)
  } catch {
    return hostStatusInvalidArgument
  }
  let sessionID = bridgeRequest.session_id

  return runOnMainThread {
    bridge.setTextInputState(
      runtimeHost: runtimeHost,
      RuntimeHostTextInputStateRequest(
        sessionID: sessionID,
        state: state
      )
    )
  }
}

/// Decode one bridge open request into one Swift text request.
private func decodeTextOpenRequest(
  _ request: DestackRustTextOpenRequest
) throws -> RuntimeHostTextInputOpenRequest {
  guard let inputType = RuntimeHostTextInputType(rawValue: Int(request.config.input_type)) else {
    throw BridgeStringError.invalidStringSlice
  }

  return RuntimeHostTextInputOpenRequest(
    configuration: RuntimeHostTextInputConfiguration(
      sessionID: request.config.session_id,
      inputType: inputType,
      isMultiline: request.config.is_multiline,
      isSecure: request.config.is_secure
    ),
    state: try decodeTextState(request.state)
  )
}

/// Decode one bridge text state into one Swift value.
private func decodeTextState(
  _ state: DestackRustTextSessionState
) throws -> RuntimeHostTextInputState {
  RuntimeHostTextInputState(
    text: try tryDecodeNativeString(state.text),
    selection: decodeTextRange(state.selection),
    composing: state.has_composing ? decodeTextRange(state.composing) : nil
  )
}

/// Decode one bridge text range into one Swift value.
private func decodeTextRange(
  _ range: DestackRustTextRange
) -> RuntimeHostTextInputRange {
  RuntimeHostTextInputRange(
    startOffset: Int(range.start_offset),
    endOffset: Int(range.end_offset)
  )
}
